File: src/main/java/hello/hellospring/service/MemberService.java
package hello.hellospring.service;

import hello.hellospring.domain.Member;
import hello.hellospring.repository.MemberRepository;
import hello.hellospring.repository.MemoryMemberRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class MemberService {
// ctrl + shift + t 를 누르면 테스트 뼈대가 자동 생성.
    private final MemberRepository memberRepository;

    @Autowired
    public MemberService(MemberRepository memberRepository) {       // dependency Injection(DI). 의존관계 주입.
        this.memberRepository = memberRepository;
    }

    /**
     * 회원 가입
     */
    public Long join(Member member) {
        // 같은 이름이 있는 중복 회원 x
//        Optional<Member> result = memberRepository.findByName(member.getName());
//        result.ifPresent(m -> {
//            throw new IllegalStateException("이미 존재하는 회원입니다.");
//        });
        // 위의 것을 줄이면 아래
//        memberRepository.findByName(member.getName())
//                .ifPresent(m -> {
//                    throw new IllegalStateException("이미 존재하는 회원입니다.");
//                });
        // 위의 것을 리팩토링하면 아래
        validateDuplicateMember(member);    // 중복 회원 검증
        memberRepository.save(member);
        return member.getId();
    }

    private void validateDuplicateMember(Member member) {
        memberRepository.findByName(member.getName())
                        .ifPresent(m -> {
                            throw new IllegalStateException("이미 존재하는 회원입니다.");
                        });
    }
    // 서비스는 비지니스를 처리하는게 서비스의 롤이므로 서비스는 비지니스에 의존적으로 설계, 용어 또한 비지니스에 맞게 선택
    // 리포지토리는 기계적으로 개발스럽게 용어를 선택

    /**
     * 전체 회원 조회
     */
    public List<Member> findMembers() {
        return memberRepository.findAll();
    }

    public Optional<Member> findOne(Long memberId) {
        return memberRepository.findById(memberId);
    }

}

File: src/main/kotlin/hello/hellospring/service/MemberService.kt
package hello.hellospring.service

import hello.hellospring.domain.Member
import hello.hellospring.dto.MemberDto
import hello.hellospring.repository.MemberRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.Optional

@Service
class MemberService @Autowired constructor(
    private val memberRepository: MemberRepository,

) {
    fun join(member: MemberDto): Long? {
        validateDuplicateMember(member)
        memberRepository.save(member)
        return member.id
    }

    fun validateDuplicateMember(member: MemberDto) {
        memberRepository.findByName(member.name)
            .ifPresent{
                throw IllegalArgumentException("이미 존재하는 회원입니다.")
            }
    }

    fun findMembers(): List<Member> {
        return memberRepository.findAll()
    }

    fun findOne(memberId: Long): Optional<Member> {
        return memberRepository.findById(memberId)
    }

}